File: schoolapp/src/services/itemLocationService.js
import apiService from './api';

const normalizeItemLocation = (loc) => {
  if (!loc || typeof loc !== 'object') return loc;

  return {
    ...loc,
    id: loc.id ?? loc.Id ?? '',
    locationName: loc.locationName ?? loc.LocationName ?? '',
    description: loc.description ?? loc.Description ?? '',
    building: loc.building ?? loc.Building ?? '',
    locationFloor: loc.locationFloor ?? loc.LocationFloor ?? '',
    locationNumber: loc.locationNumber ?? loc.LocationNumber ?? null,
    capacity: loc.capacity ?? loc.Capacity ?? null,
    isActive: loc.isActive ?? loc.IsActive,
    companyId: loc.companyId ?? loc.CompanyId ?? '',
    schoolId: loc.schoolId ?? loc.SchoolId ?? '',
    createdDate: loc.createdDate ?? loc.CreatedDate ?? null,
    status: loc.status ?? loc.Status ?? '',
    statusMessage: loc.statusMessage ?? loc.StatusMessage ?? '',
  };
};

const normalizeItemLocationResponse = (response) => {
  if (Array.isArray(response)) return response.map(normalizeItemLocation);
  if (response?.data && Array.isArray(response.data)) return response.data.map(normalizeItemLocation);
  if (response?.items && Array.isArray(response.items)) return response.items.map(normalizeItemLocation);
  return normalizeItemLocation(response);
};

export const itemLocationService = {
  async getAll() {
    try {
      const response = await apiService.get('/itemlocation');
      return normalizeItemLocationResponse(response);
    } catch (error) {
      console.error('Item Location API error:', error);
      throw new Error(error.message || 'Failed to fetch item locations');
    }
  },

  async getById(id) {
    try {
      const response = await apiService.get(`/itemlocation/${id}`);
      return normalizeItemLocation(response);
    } catch (error) {
      throw new Error(error.message || 'Failed to fetch item location');
    }
  },

  async create(data) {
    try {
      const response = await apiService.post('/itemlocation', data);
      return normalizeItemLocation(response);
    } catch (error) {
      throw new Error(error.message || 'Failed to create item location');
    }
  },

  async update(id, data) {
    try {
      const response = await apiService.put(`/itemlocation/${id}`, data);
      return normalizeItemLocation(response);
    } catch (error) {
      throw new Error(error.message || 'Failed to update item location');
    }
  },

  async delete(id) {
    try {
      await apiService.delete(`/itemlocation/${id}`);
      return true;
    } catch (error) {
      throw new Error(error.message || 'Failed to delete item location');
    }
  }
};
export default itemLocationService;

File: SchoolDemo.Application/Services/ItemLocationService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using SchoolDemo.Domain.Entities;
using SchoolDemo.Domain.Interfaces;

namespace SchoolDemo.Application.Services;

public class ItemLocationService : IItemLocationService
{
    private readonly IItemLocationRepository _repository;

    public ItemLocationService(IItemLocationRepository repository)
    {
        _repository = repository;
    }

    public async Task<ItemLocationResponse?> GetByIdAsync(Guid id)
    {
        var itemLocation = await _repository.GetByIdAsync(id);
        return itemLocation == null ? null : MapToResponse(itemLocation);
    }

    public async Task<IEnumerable<ItemLocationResponse>> GetAllAsync()
    {
        var itemLocations = await _repository.GetAllAsync();
        return itemLocations.Select(MapToResponse);
    }

    public async Task<ItemLocationResponse> CreateAsync(ItemLocationRequest request)
    {
        var itemLocation = new ItemLocation
        {
            Id = Guid.NewGuid(),
            LocationName = request.LocationName,
            Description = request.Description,
            Building = request.Building,
            LocationFloor = request.LocationFloor,
            LocationNumber = request.LocationNumber,
            Capacity = request.Capacity,
            IsActive = request.IsActive ?? true,
            CompanyId = request.CompanyId,
            SchoolId = request.SchoolId,
            IsDeleted = false,
            CreatedBy = Guid.Empty, // Resolved in repository
            CreatedDate = DateTime.UtcNow,
            ModifiedDate = DateTime.UtcNow,
            Status = "Active",
            StatusMessage = "Item location created successfully"
        };

        var created = await _repository.AddAsync(itemLocation);
        return MapToResponse(created);
    }

    public async Task<ItemLocationResponse?> UpdateAsync(Guid id, ItemLocationRequest request)
    {
        var existing = await _repository.GetByIdAsync(id);
        if (existing == null || existing.IsDeleted == true)
        {
            return null;
        }

        existing.LocationName = request.LocationName ?? existing.LocationName;
        existing.Description = request.Description ?? existing.Description;
        existing.Building = request.Building ?? existing.Building;
        existing.LocationFloor = request.LocationFloor ?? existing.LocationFloor;
        existing.LocationNumber = request.LocationNumber ?? existing.LocationNumber;
        existing.Capacity = request.Capacity ?? existing.Capacity;
        existing.IsActive = request.IsActive ?? existing.IsActive;
        existing.CompanyId = request.CompanyId != Guid.Empty ? request.CompanyId : existing.CompanyId;
        existing.SchoolId = request.SchoolId != Guid.Empty ? request.SchoolId : existing.SchoolId;
        existing.ModifiedDate = DateTime.UtcNow;
        existing.Status = "Updated";
        existing.StatusMessage = "Item location updated successfully";

        var updated = await _repository.UpdateAsync(existing);
        return MapToResponse(updated);
    }

    public async Task<bool> DeleteAsync(Guid id)
    {
        var existing = await _repository.GetByIdAsync(id);
        if (existing == null || existing.IsDeleted == true)
        {
            return false;
        }

        await _repository.DeleteAsync(id);
        return true;
    }

    private static ItemLocationResponse MapToResponse(ItemLocation itemLocation)
    {
        return new ItemLocationResponse
        {
            Id = itemLocation.Id,
            LocationName = itemLocation.LocationName,
            Description = itemLocation.Description,
            Building = itemLocation.Building,
            LocationFloor = itemLocation.LocationFloor,
            LocationNumber = itemLocation.LocationNumber,
            Capacity = itemLocation.Capacity,
            IsActive = itemLocation.IsActive,
            CompanyId = itemLocation.CompanyId,
            SchoolId = itemLocation.SchoolId,
            CreatedBy = itemLocation.CreatedBy,
            CreatedDate = itemLocation.CreatedDate,
            ModifiedBy = itemLocation.ModifiedBy,
            ModifiedDate = itemLocation.ModifiedDate,
            IsDeleted = itemLocation.IsDeleted,
            Status = itemLocation.Status,
            StatusMessage = itemLocation.StatusMessage
        };
    }
}
